### PR TITLE
riotboot: set FLASHFILE to RIOTBOOT_EXTENDED_BIN if riotboot feature is used

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

Previously, even an application that had "FEATURES_REQUIRED += riotboot"
set would still flash the non-riotboot binary on "make flash".
This is usually not what the user wants.

This commit set's the FLASHFILE variable to the combined "riotboot
bootloader + slot0 + empty slot1" binary, if the riotboot feature is used.
This has the effect that "make all", "make flash" and "make flash-only" will compile and/or flash a working riotboot setup.

This was previously done for tests/riotboot within that application's Makefile. This PR moves the definition into ```makefiles/boot/riotboot.inc.mk```, guarded by whether the riotboot feature is used.

### Testing procedure

- confirm that non-riotboot flashing still works (e.g., by flashing hello-world)
- run "make flash" in tests/riotboot. It should flash the combined binary

### CODE REVIEW

**Comment by kaspar030** [comment]
- fixed some issues @cladmi pointed out IRL

**Comment by cladmi** [comment]
"riotboot: use riotboot if feature is used" The PR title and commit message should maybe be more precise and say "use the riotboot extended firmware" or any other descriptive name.

This was not done at the beginning as `FLASHFILE` was not supported for all boards and required hacks. Doing it globally before it was supported could have led to overwriting both `ELFFILE`, `HEXFILE`, `BINFILE` and cause recursive definitions issues

Now it should be safe to do.

**Comment by kaspar030** [comment]
> "riotboot: use riotboot if feature is used" The PR title and commit message should maybe be more precise

agreed. Changed to what the commit actually does.

**Comment by kaspar030** [comment]
@cladmi could you take another look?

**Comment by cladmi** [comment]
This is currently changing multiple things in the same PR without doing them completely and giving testing for it.

This changed without justifying the firmware flashed for `tests/riotboot`.
Even if it is a needed change it is not explained why.

The documentation still recommends to use `riotboot/flash-combined-slot0` for flashing instead of using `flash` directly.

The documentation update for the previous APP_VER api change is not complete as many examples still say to use `APP_VER=$(date +%s)`.

**Comment by kaspar030** [comment]
> This changed without justifying the firmware flashed for `tests/riotboot`.
> Even if it is a needed change it is not explained why.

- added a paragraph to the commit description

> The documentation still recommends to use `riotboot/flash-combined-slot0` for flashing instead of using `flash` directly.

... which is not wrong. Anyhow, changed it to say "make flash".

> The documentation update for the previous APP_VER api change is not complete as many examples still say to use `APP_VER=$(date +%s)`.

Which is also not wrong. I piggybacked the fix for actually wrong information, I can kick it out if you insist. I do not want to have more unrelated fixes in here.

**Comment by kaspar030** [comment]
@cladmi are your comments sufficiently addressed?

File: makefiles/boot/riotboot.mk [hunk 1]
```diff
@@ -126,6 +126,13 @@ riotboot/slot1: $(SLOT1_RIOT_BIN)
 # Default flashing rule for bootloader + slot 0
 riotboot/flash: riotboot/flash-slot0 riotboot/flash-bootloader
 
+# make applications that use the riotboot feature default to actually using it
+# Target 'all' will generate the combined file directly.
+# It also makes 'flash' and 'flash-only' work without specific command.
+ifneq (,$(filter riotboot, $(FEATURES_USED)))
```
Comment (cladmi) [inline makefiles/boot/riotboot.mk]: This guard is duplicating the one from the file. So is not necessary.

Comment (kaspar030) [reply makefiles/boot/riotboot.mk]: yup, fixed

**Comment by kaspar030** [comment]
> My only concern is that now there is no way of flashing just the application without riotboot, right?

That might indeed be an issue. I'll try to come up with something.

**Comment by kaspar030** [comment]
> My only concern is that now there is no way of flashing just the application without riotboot, right?

Do you mean e.g., only flash slot0 or slot1? That's still possible using ```make riotboot/flash-slotX```.
Or do you mean flashing a non-riotboot version?

**Comment by fjmolinas** [comment]
> Do you mean e.g., only flash slot0 or slot1? That's still possible using make riotboot/flash-slotX.
> Or do you mean flashing a non-riotboot version?

Non-riotboot version.

**Comment by kaspar030** [comment]
> Non-riotboot version.

Ok, this would require disabling usage of an available and optional or required feature. It would probably need a new variable like "FEATURES_DISABLED", which will make people scream "YOU CANNOT DISABLE WHAT THE HARDWARE ALWAYS PROVIDES!".

Unless we had a strong use-case that is now not supported anymore, I suggest we postpone this.

**Comment by fjmolinas** [comment]
> Ok, this would require disabling usage of an available and optional or required feature. It would probably need a new variable like "FEATURES_DISABLED", which will make people scream "YOU CANNOT DISABLE WHAT THE HARDWARE ALWAYS PROVIDES!".

True.

> Unless we had a strong use-case that is now not supported anymore, I suggest we postpone this.

Ok, I agree. @cladmi do you still have issues with this?

**Comment by fjmolinas** [comment]
I'm OK with the changes, but since @cladmi requested most of them I would like for him to ACK that his concerns are addressed.

**Comment by kaspar030** [comment]
@cladmi ping

**Comment by kaspar030** [comment]
I realized that "bootloaders/riotboot/README.md" also had outdated testing instructions. I've replaced them with a pointer to "tests/riotboot/README.md".

I think this is good to go, could someone take a look?

**Comment by fjmolinas** [comment]
@cladmi have your issues been addressed?

File: bootloaders/riotboot/README.md [hunk 2]
```diff
@@ -94,17 +100,9 @@ Dedicated make targets are available to build and flash several slots:
...
-
-`BOARD=samr21-xpro APP_VER=$(date +%s) make -C tests/riotboot/ riotboot/flash-combined-slot0 test`
+slot 0 being valid, thus being booted). This is done automatically by `make
+flash` if the `riotboot` feature is used.
 
-For the second slot:
+## Testing riotboot
 
-`BOARD=samr21-xpro APP_VER=$(date +%s) make -C tests/riotboot/ riotboot/flash-slot1 test`
+See tests/riotboot/README.md.
```
Comment (fjmolinas) [inline bootloaders/riotboot/README.md]: Can you put this as a link? [tests/riotboot/README.md](https://github.com/RIOT-OS/RIOT/blob/master/tests/riotboot/README.md)

Comment (kaspar030) [reply bootloaders/riotboot/README.md]: good idea, fixed

**Comment by kaspar030** [comment]
Two more boards got riotboot support, but are too small to fit the test in half the flash. Pushed a commit adding them to the blacklist.

**Comment by cladmi** [comment]
Comments were addressed.

**Comment by fjmolinas** [comment]
GO!

**Pull request merged**